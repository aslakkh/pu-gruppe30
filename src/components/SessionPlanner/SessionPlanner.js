import React, { Component } from 'react'
import {FormGroup, FormControl, ControlLabel, Button} from 'react-bootstrap'
import { planSession } from '../../helpers/auth'
import './SessionPlanner.css'

/*
TODO:
- prevent from moving back to homepage on submit
- Add better feedback to user (about validation state)

*/

var DatePicker = require("react-bootstrap-date-picker");

export default class SessionPlanner extends Component{

    constructor(props){
        super(props);
        //this.getInitialState = this.getInitialState.bind(this);
        this.handleDateChange = this.handleDateChange.bind(this);
        this.handleGoalChange = this.handleGoalChange.bind(this);
        this.handleClick = this.handleClick.bind(this);
        this.validateState = this.validateState.bind(this);
        this.state = {
            date: new Date(new Date().getTime() + (24*60*60*1000)).toISOString(),
            goal: "",
            course: this.props.course,
        }
    }

    /*
    getInitialState(){
        var d = new Date(new Date().getTime() + (24*60*60*1000)).toISOString();
        return{
            date: d
        }
    }*/

    //handler for change in datepicker. Sets state.date to the ISO-string provided by datepicker
    handleDateChange(e){
        
        this.setState({
            date: e,
        });
    }

    //handler for change in text area
    handleGoalChange(e){
        this.setState({
            goal: e.target.value,
        });
    }

    //handles click on submit button
    handleClick(e){
        e.preventDefault(); //prevent default HTML-action on submitting a form
        let d = new Date(this.state.date); //creates a date object from the ISO-string provided by datepicker
        if(this.validateState()){
            //sets the time of the date sent to firebase to system time
            //so that every planned session gets a unique key (in ms)
            let today = new Date();
            d.setHours(today.getHours(), today.getMinutes(), today.getSeconds(), today.getMilliseconds());
            planSession(this.state.course, d.getTime(), this.state.goal);
        }
        else{
            //TODO: provide better user feedback
            console.log("Error: tried submitting invalid values");
        }
        
    }

    validateState(){
        //validates that date is after today
        //validates that goal is not empty
        let today = new Date(new Date);
        let stateDate = new Date(this.state.date);
        stateDate.setHours(0,0,0,0);
        return(stateDate.getTime() >= today.getTime() && this.state.goal.trim() !== "")
    }

    componentDidMount(){
        //console.log(this.state.course);
    }


    //Component renders a Formgroup containing a textarea and a datepicker, and a button for submitting data
    render(){
        //<DatePicker defaultValue={this.state.date}/>
        return(
            <div>
                <FormGroup 
                controlId="planSessionForm"
                >
                    <ControlLabel>Goal/Description</ControlLabel>
                    <FormControl 
                    componentClass="textarea" 
                    placeholder="Write what you plan to do here."
                    type="text" 
                    value={this.state.goal}
                    onChange={this.handleGoalChange}
                    />
                    <ControlLabel>Date</ControlLabel>
                    <DatePicker id="example-datepicker" value={this.state.date} onChange={this.handleDateChange}/>
                </FormGroup>
                <Button
                bsStyle='primary'
                bsSize='large'
                onClick={this.handleClick}
                >
                Plan this session</Button>
            </div>
        );
        
    }
}