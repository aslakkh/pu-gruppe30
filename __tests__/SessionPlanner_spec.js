/**
 * Created by jan on 20/04/2017.
 */
import React from 'react';
import renderer from 'react-test-renderer';
import SessionPlanner from '../src/components/SessionPlanner/SessionPlanner.js';
import TestUtils from "react-addons-test-utils"

jest.useFakeTimers();
const DATE_TO_USE = new Date('2016');
const _Date = Date;
global.Date = jest.fn(() => DATE_TO_USE);
global.Date.UTC = _Date.UTC;
global.Date.parse = _Date.parse;
global.Date.now = _Date.now;
const courses={
    TES1000:{active: true,
        goals: {dailyGoal:{active: false,
            timeSet: 0,
            timeSpent: 0,
            value: 0},
            monthlyGoal:
                {active:false,
                    timeSet:0,
                    timeSpent:0,
                    value:0},
            weeklyGoal:
                {active:false,
                    timeSet:0,
                    timeSpent:0,
                    value:0}

        }
    }};
const coursesNull=null;
describe("AddSession", function () {
    let renderer2;
    let component2;
    jest.useFakeTimers();
    Date.getTime = jest.fn(() => 1482363367071);
    Date.now = jest.fn(() => 1482363367071);
    renderer2 = TestUtils.createRenderer();
    renderer2.render(
        <SessionPlanner course={'TES1000'}/>
    );
    component2 = renderer2.getRenderOutput();
    it("One course", () => {
        expect(component2).toMatchSnapshot();
    });
});