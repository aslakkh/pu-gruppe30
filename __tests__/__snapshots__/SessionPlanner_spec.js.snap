// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddSession One course 1`] = `
<div>
  <FormGroup
    bsClass="form-group"
    controlId="planSessionForm"
    validationState="error"
  >
    <ControlLabel
      bsClass="control-label"
      srOnly={false}
    >
      Goal/Description
    </ControlLabel>
    <FormControl
      bsClass="form-control"
      componentClass="textarea"
      onChange={[Function]}
      placeholder="Write what you plan to do here."
      type="text"
      value=""
    />
  </FormGroup>
  <FormGroup
    bsClass="form-group"
    controlId="datePicker"
    validationState="error"
  >
    <ControlLabel
      bsClass="control-label"
      srOnly={false}
    >
      Date
    </ControlLabel>
    <DatePicker
      autoFocus={false}
      calendarPlacement="bottom"
      cellPadding="5px"
      clearButtonElement="Ã—"
      dateFormat="MM/DD/YYYY"
      dayLabels={
        Array [
          "Sun",
          "Mon",
          "Tue",
          "Wed",
          "Thu",
          "Fri",
          "Sat",
        ]
      }
      disabled={false}
      id="example-datepicker"
      instanceCount={0}
      monthLabels={
        Array [
          "January",
          "February",
          "March",
          "April",
          "May",
          "June",
          "July",
          "August",
          "September",
          "October",
          "November",
          "December",
        ]
      }
      nextButtonElement=">"
      onChange={[Function]}
      previousButtonElement="<"
      showClearButton={true}
      showTodayButton={false}
      style={
        Object {
          "width": "100%",
        }
      }
      todayButtonLabel="Today"
      value="2016-01-01T00:00:00.000Z"
    />
  </FormGroup>
  <Button
    active={false}
    block={false}
    bsClass="btn"
    bsSize="large"
    bsStyle="primary"
    disabled={false}
    onClick={[Function]}
  >
    Plan this session
  </Button>
</div>
`;
